packages/block-editor/src/utils/transform-styles/index.js

```diff
@@ -1,39 +1,132 @@
+	{ type: 'pseudo-class', content: ':root' },
+	{ type: 'pseudo-class', content: ':where(body)' },
+	{ type: 'pseudo-class', content: ':where(:root)' },
+	{ type: 'pseudo-class', content: ':where(html)' },
+];
+
+/**
+ * Prefixes root selectors in a way that ensures consistent specificity.
+ * This requires special handling, since prefixing a classname before
+ * html, body, or :root will generally result in an invalid selector.
+ *
+ * Some libraries will simply replace the root selector with the prefix
+ * instead, but this results in inconsistent specificity.
+ *
+ * This function instead inserts the prefix after the root tags but before
+ * any other part of the selector.
+ *
+ * @param {string} prefix   The prefix.
+ * @param {string} selector The selector.
+ *
+ * @return {string} The prefixed root selector.
+ */
+function prefixRootSelector( prefix, selector ) {
+	// Use a tokenizer, since regular expressions are unreliable.
+	const tokenized = parsel.tokenize( selector );
+
+	// Find the last token that contains a root selector by walking back
+	// through the tokens.
+	const lastRootIndex = tokenized.findLastIndex( ( { content, type } ) => {
+		return ROOT_SELECTOR_TOKENS.some(
+			( rootSelector ) =>
+				content === rootSelector.content && type === rootSelector.type
+		);
+	} );
+
+	// Walk forwards to find the combinator after the last root.
+	// This is where the root ends and the rest of the selector begins,
+	// and the index to insert before.
+	// Doing it this way takes into account that a root selector like
+	// 'body' may have additional id/class/pseudo-class/attribute-selector
+	// parts chained to it, which is difficult to quantify using a regex.
+	const insertionPoint = tokenized.findIndex( ( { type }, index ) => {
+		return index > lastRootIndex && type === 'combinator';
+	} );
+
+	// Tokenize and insert the prefix with a ' ' combinator before it.
+	const tokenizedPrefix = parsel.tokenize( prefix );
+	tokenized.splice(
+		// Insert at the insertion point, or the end.
+		insertionPoint === -1 ? tokenized.length : insertionPoint,
+		0,
+		{
+			type: 'combinator',
+			content: ' ',
+		},
+		...tokenizedPrefix
+	);
+
+	return parsel.stringify( tokenized );
+}
+
 function transformStyle(
 	{ css, ignoredSelectors = [], baseURL },
-	wrapperSelector = ''
+	wrapperSelector = '',
+	transformOptions
 ) {
-	// When there is no wrapper selector or base URL, there is no need
+	// When there is no wrapper selector and no base URL, there is no need
 	// to transform the CSS. This is most cases because in the default
 	// iframed editor, no wrapping is needed, and not many styles
 	// provide a base URL.
 	if ( ! wrapperSelector && ! baseURL ) {
 		return css;
 	}
-	const postcssFriendlyCSS = css
-		.replace( /:root :where\(body\)/g, 'body' )
-		.replace( /:where\(body\)/g, 'body' );
+
 	try {
+		const excludedSelectors = [
+			...ignoredSelectors,
+			...( transformOptions?.ignoredSelectors ?? [] ),
+			wrapperSelector,
```

Comment:
Do we need to escape normal CSS selectors here e.g. `:where(...)` for use as a regex?

Reply:
The `match` method works for strings as well as regexes, so I think it should be ok.

Reply:
Did you apply the diff in my review? 

That shows that if the `wrapperSelector` is `:where(.my-namespace)` and the selector to be prefixed contains `:where(.my-namespace)` it still prefixes when it shouldn't.

<img width="835" alt="Screenshot 2024-08-26 at 6 51 16 PM" src="https://github.com/user-attachments/assets/001c443b-8260-4d64-bfa1-aa9a68927022">

Reply:
I'll take a look and make sure there's a unit test 👍

Reply:
That should be fixed now in the latest commit, thanks for catching it 👍

Reply:
Works well now 🚀 

I like the simple check for whether the excluded selector is a RegExp instance. Means all kinds of funky CSS selectors are covered.